### ISSUE TITLE
forkIfNeeded can kill the calling thread

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#5723]
## Environment

* SuperCollider version: Current, for years now
* Operating system: Any
* Other details (Qt version, audio driver, etc.): n/a

## Steps to reproduce

```supercollider
(
s.waitForBoot {
	b = Buffer.alloc(s, 4096, 1, { |buf| buf.sine1Msg([1]) });
};
)

(
fork {
	var cond = CondVar.new;
	var data;
	b.getToFloatArray(wait: -1, action: { |array|
		"called action".debug;
		data = array;
		cond.signalOne;
	});
	cond.wait;
	data.postln;  // We never get here :-O
}
)
```

Printed:

```
-> a Routine
called action
```

`data.postln` never ran, and everything following it will also fail to execute.

## Expected vs. actual behavior

We expect to be able to block an asynchronous operation using CondVar (here, I'm trying to help with the transition away from Condition, but Condition is affected as well -- in the following, please read `CondVar` as referring to any of the thread-sync classes).

If this asynchronous operation itself uses CondVar (including indirectly by `server.sync`) to block for sub-operations *and* it uses `forkIfNeeded`, then it effectively hijacks the calling thread and when this operation comes to an end, the calling thread can't be resumed.

I've run into this behavior probably dozens of times over the last several years.

I thought perhaps that `forkIfNeeded` should yield something to put the thread into paused state rather than stopped:

```
	forkIfNeeded { arg clock, quant, stackSize;
		if(thisThread.isKindOf(Routine)) {
			this.value;
			\keepAlive.debug("yielding").yield;
		} {
			^this.fork(clock, quant, stackSize)
		};
		// I suspect this is a nonsense return?
		^thisThread;
	}
```

... but this doesn't fix the problem.

I also tried to post something when a routine's function ends:

```
	prStart { arg inval;
		">> Routine:prStart".debug;
		func.value(inval);
		"<< Routine:prStart".debug;
		// if the user's function returns then always yield nil
		nil.alwaysYield;
	}
```

But I get:

```
-> a Routine
>> Routine:prStart
called action
```

So the thread is dying but not because the function returned.

I don't know exactly why the thread is dying.

So I think the paths forward are:

- Option A: Someone figures out why the thread is ending, and fixes it in `forkIfNeeded`.
- Option B: `forkIfNeeded` is fatally flawed, should be deprecated, and its uses in the class library replaced with a simple `fork`. (If `forkIfNeeded` is called from a routine, then *it doesn't own the thread* and it doesn't have any right to do anything to the thread's flow. So I think there's a good case to be made that it was introduced without understanding all of its side effects, and that could be un-done by just removing it.)

Open to other options.

Side note: If, in cases such as `getToFloatArray`, we want to have looping and blocking, *recursion* can handle the looping without forking another thread -- proof of concept below. The recursive call is happening in an async responder, after `readChunkAndContinue` already exited. So the stack will not grow, and this can loop as long as necessary.

```
(
var readChunkAndContinue = { |buffer, array, frame, numFrames, arrayPos, chunkSize, action|
	buffer.getn(frame, min(chunkSize, numFrames - arrayPos), { |data|
		data.size.debug("received size");
		array.overWrite(data.as(FloatArray), arrayPos);
		arrayPos = arrayPos + data.size;
		if(arrayPos < numFrames) {
			readChunkAndContinue.(buffer, array,
				frame + data.size, numFrames, arrayPos,
				chunkSize, action
			)
		} { action.value(array) };
	});
};

f = { |buffer, startFrame = 0, numFrames = -1, action|
	var chunkSize = 1633;  // imitate the method
	var result;
	
	if(numFrames < 0) { numFrames = buffer.numFrames };
	result = FloatArray.newClear(numFrames);
	
	readChunkAndContinue.(buffer, result, startFrame, numFrames, 0, chunkSize, action);
};
)

(
fork {
	var cond = CondVar.new;
	var data;
	f.(b, action: { |array|
		"called action".debug;
		data = array;
		cond.signalOne;
	});
	cond.wait;
	data.postln;  // OK!
}
)

received size: 1633
received size: 1633
received size: 830
called action
FloatArray[ -0.0030679572373629, 0.0030679572373629, ...  // yep!
```

**Comment by Spacechild1** [comment]
[off-topic]

> Side note: If, in cases such as getToFloatArray, we want to have looping and blocking, recursion can handle the looping without forking another thread -- proof of concept below. 

One reason why `getToFloatArray` uses fork + wait is that the user can control the wait time between chunk transfers.

**Comment by jamshark70** [comment]
> One reason why getToFloatArray uses fork + wait is that the user can control the wait time between chunk transfers.

Of course this can be handled with recursion as well.

The proof-of-concept places the recursive call in an async action function.

But the recursive call could just as easily be scheduled for the user's desired wait time, and the difference in the recursive behavior would be... no difference at all.

So it's a non-objection. The desire to let the user control the wait time does not rule out a recursive approach.

**Comment by Spacechild1** [comment]
You're right, thanks.

**Comment by dyfer** [comment]
Thanks for the report.

My 2¢: 
> Option B: forkIfNeeded is fatally flawed, should be deprecated, and its uses in the class library replaced with a simple fork

I find `forkIfNeeded` quite useful in writing modular code that can be run on its own, as well as embedded in calling routines, while maintaining proper order of operations. If `forkIfNeeded` is to be deprecated, I think it would be imperative to implement a comparable solution... which, if possible (hopefully!), would probably mean that fixing `forkIfNeeded` is the way to go.

**Comment by jamshark70** [comment]
OK, I see why it's happening.

I added some debugging posts into Object:yield, and the initial reproducer prints the following:

```
before yield state, value: [ 3, hang ]
state after yield: 5
before yield state, value: [ 3, hang ]
state after yield: 5
before yield state, value: [ 3, hang ]
state after yield: 5
called action
before yield state, value: [ 3, wait ]
state after yield: 5
```

It's lucky for troubleshooting that `getToFloatArray` is still using the old Condition, which yields `\hang` when blocking, while my test case uses CondVar (yields `\wait`). So the order is clear.

The outer `cond.wait` is taking place after `getToFloatArray` has completely finished, and already fired its action. Because the action has already evaluated, there is nothing remaining to do `cond.signalOne` -- there's no signal, thread doesn't resume.

`getToFloatArray` with a negative wait time, then, behaves as if it's synchronous -- but, with a positive wait time, it's asynchronous and requires a condition for blocking. Compare:

```
// wait: -1, remove `cond.wait`
(
fork {
	var cond = CondVar.new;
	var data;
	b.getToFloatArray(wait: -1, action: { |array|
		"called action".debug;
		data = array;
		cond.signalOne;
	});
	// cond.wait;  -- nope, skip this
	data.postln;  // We never get here :-O
}
)

before yield state, value: [ 3, hang ]  -- 3x
state after yield: 5  -- 3x
called action
FloatArray[ -0.0030679572373629, 0.0030679572373629, ... ]

// remove *both* `wait: -1` and `cond.wait`
before yield state, value: [ 3, 0.01 ]  -- 3x
state after yield: 5  -- 3x
nil  -- whaaa? data aren't available
before alwaysYield state, value: [ 3, nil ]
state after alwaysYield: 6
called action  -- oh... because action got called after thread continued
```

So the flow of control is either synchronous or asynchronous, depending on whether the method is blocking for a reply or scheduling forward.

That makes me nervous... I like rules that are simple to explain to new users. "It's waiting for server replies, so you should block the thread" is a simple rule. "It's waiting for server replies, and you need to know the method's implementation details so that you can either block or not block depending on whether the method is blocking" is not a simple rule (and breaks abstraction -- the user shouldn't have to care if the method is blocking or not).

This characteristic of `forkIfNeeded` -- here it's synchronous, there it's not -- makes me skeptical and I expect I'll avoid it from here on out. But I can go along with @dyfer's point that it can be useful in some situations.

So perhaps an option C could be to review uses of forkIfNeeded in the main class library and replace it with `fork` in any situation where there's a `.sync` or Condition/CondVar. *And* document carefully the cases where `forkIfNeeded` is unsafe.

- SoundFile `asBuffer` (for remote servers)
- Buffer `*sendCollection`, `streamCollection`, `loadToFloatArray`, `getToFloatArray`
- BundleNetAddr `closeBundle` (but this one is not really user facing)

^^ These all use `.sync` inside a `forkIfNeeded`.

- UnitTest `*runAll`, `runTestMethod`, `*prRunAllTestMethods`, `*prRunWithinSetUpClass`, `runScript` -- I'm not sure about these. These are running code in unit test classes, over which `UnitTest` itself has no control. So they might be OK but we have no guarantee of that.
- Server:prWaitForPidRelease -- not user-facing, probably OK.
- Quarks `*load` and `*checkForUpdates` -- probably OK.

I recently encountered this in the VSTPlugin extension as well.

**Comment by smrg-lm** [comment]
> replace it with `fork`

That could work just fine. The only use case that will break is when forkIfNeeded is called from the main thread and anything already running on it should be async code anyway.

**Comment by smrg-lm** [comment]
I mean, library wise, user code may break because the behavior was changed and it manifest a different bug. But it's better to remove because it also hides the logic of the code imo.

**Comment by Spacechild1** [comment]
Now that I had a closer look, I realize that the whole issue is deeply tied to the semantic difference between `Condition` + `CondVar`:

```
(
{
	var cond = Condition();
	cond.test = true;
	cond.signal;
	"wait".postln;
	cond.wait;
	"done".postln; // ok
}.fork
)

(
{
	var cond = CondVar();
	cond.signalOne;
	"wait".postln;
	cond.wait;
	"done".postln; // never reached
}.fork
)
```

`Condition` has a similar behavior as a semaphore or Win32 Event: if you signal it before calling `wait` it won't block.

`CondVar`, on the other hand, behaves like a pthread condition variable: if you signal it before calling `wait`, the signal is lost.

---

If you replace `CondVar` with `Condition` in the original code example, it works as expected:

```
(
s.waitForBoot {
	b = Buffer.alloc(s, 4096, 1, { |buf| buf.sine1Msg([1]) });
};
)

(
fork {
	var cond = Condition.new;
	var data;
	b.getToFloatArray(wait: -1, action: { |array|
		"called action".debug;
		data = array;
		cond.test = true;
		cond.signal;
	});
	cond.wait;
	data.postln;  // We never get here :-O
}
)
```

AFAICT, `CondVar` is a low level threading primitive and it comes with some caveats. I'm not sure it's really the right tool for the job here. At least there should be disclaimer in the help file! @jamshark70 What's your reason for using `CondVar` instead of `Condition`?

As a side note: AFAICT there is no formal guarantee that an async method call *has* to be executed asynchronously. There is no reason why the action function shouldn't be called synchronously under certain circumstances.

---

That being said, I agree that `forkIfNeeded` inside an async method is problematic because the method will essentially block when executed inside a Routine. I don't think this is what people want.

Of course, some users might (implicitly) rely on the current behavior, so replacing `forkIfNeeded` with `fork` might break user code. On the other hand, one really shouldn't expect async functions to be blocking, so this would hardly count as a regression. Still we should think carefully about it.

**Comment by Spacechild1** [comment]
TL;DR: `forkIfNeeded` is not broken, but it can change the semantic of async method calls. This can have unexpected consequences when people try to synchronize with `CondVar`.

The documentation of `CondVar` needs to be improved and the difference to `Condition` has to be stated prominently!

**Comment by Spacechild1** [comment]
One more thing about `CondVar`: like a posix condition variable, it is not supposed to be used alone, it should always be used together with some kind of predicate.

In the original code, the predicate would be the value of `data`. This means it should have been written like this:
```
(
s.waitForBoot {
	b = Buffer.alloc(s, 4096, 1, { |buf| buf.sine1Msg([1]) });
};
)

(
fork {
	var cond = CondVar.new;
	var data;
	b.getToFloatArray(wait: -1, action: { |array|
		"called action".debug;
		data = array;
		cond.signalOne;
	});
	cond.wait({ data.notNil }); // don't forget the predicate!
	data.postln;
}
)
```

**Comment by Spacechild1** [comment]
TL;DR: `forkIfNeeded` is not broken, but it can change the behavior of async method calls. This can have unexpected consequences when people try to synchronize with `CondVar` without a proper predicate.

The documentation of `CondVar` needs to be improved and the difference to `Condition` has to be stated prominently!

**Comment by alln4tural** [comment]
I have a function i use to load some buffers.It's implemented with forkIfNeeded because sometimes i call it manually, sometimes it's called from other functions that have already forked.This is my use case for forkIfNeeded, and i think it would break if you "replace it with fork", e.g. my calling functions WOD no longer wait for the buffers to load.I now see @dyfer stated this more formally/elegantly aboveCheers,eddi--https://alln4tural.bandcamp.comOn 12/02/2022, 11:32 Lucas Samaruga ***@***.***> wrote:

   
  I mean, library wise, user code may break because the behavior was changed and it manifest a different bug. But it's better to remove because it also hides the logic of the code imo.

**Comment by Spacechild1** [comment]
@alln4tural I think nobody wants to replace `forkIfNeeded` *itself*  with `fork`. It's really about the use of `forkIfNeeded` in some methods in the Class Library. You can absolutely continue to use it in your functions.

> This is my use case for forkIfNeeded, and i think it would break if you "replace it with fork", e.g. my calling functions would no longer wait for the buffers to load.

Can you show a code example? It's not immediately clear to me how forking would break code.

**Comment by smrg-lm** [comment]
@Spacechild1 I see it will break if user code relies on it to block inside a routine, which was the behavior so far for `getToFloatArray` and other user methods of buffer. This is the classical "is not a bug, it's a feature" situation. There is no way to solve it without breaking user code.

**Comment by Spacechild1** [comment]
> if user code relies on it to block inside a routine

I see. Ouch.

> This is the classical "is not a bug, it's a feature" situation.

Also know as: "someone's bug is someone else's feature"

**Comment by jamshark70** [comment]
TL;DR OK, I understand -- it's my incorrect usage, to rely on predicate-less syncing. I won't argue for code changes, but this needs more documentation.

> Condition has a similar behavior as a semaphore or Win32 Event: if you signal it before calling wait, it won't block.

Almost... if you satisfy the condition and signal it before calling wait, it won't block.

Condition has `hang` and `unhang`, which bypass the Boolean. Perhaps this was always a mistake and those never should have been added, because it's impossible to use those methods and be aware of a state change preceding `hang`. I guess CondVar allows usage without a predicate to match the `hang` / `unhang` interface (and then replicates the problem with Condition `hang` / `unhang`).

So the problem isn't semaphore vs pthread semantics. It's that using either class without a predicate is risky. (Your Condition example does `test_(true)` so it is not *only* replacing the class.)

Unfortunately, this is common practice and it will take some effort to change the culture around that. I can see the point that syncing without a predicate was not a best practice (just one more case where taking a shortcut eventually causes problems) and I can go through and check my code later.

> AFAICT, CondVar is a low level threading primitive and it comes with some caveats. I'm not sure it's really the right tool for the job here. @jamshark70 What's your reason for using CondVar instead of Condition?

Condition is older. At one point, we thought it supported timeouts but `cond.wait(number)` doesn't work correctly if the condition is met before the timeout expires. Then I figured out a way to have a properly-functioning timeout in Condition. But virtual_dog (who has since left the project, so I won't `@` them) found a number of conceptual and API problems with that and concluded that the best way to fix them was to add a new class, CondVar.

I've been switching over to use CondVar wherever possible on the assumption that CondVar's design addresses problems in Condition, hence, is better.

Also I think we shouldn't confuse users with two very similar sync classes. We should pick one and standardize on it (and, if CondVar is better, then why would people continue to use Condition?).

Basically the documentation concern is to establish a standard practice for thread sync. Users face a lot of guesswork on this topic currently.

> Actually, I have such a function in my code base: VSTPlugin.prQuery checks if the plugin info is already available on the Client. If yes, it calls the action function immediately; if no, it requests the info from the Server and calls the action function when the Server reply arrives.

The specific problem I had, which set all of this off, is to wait until the plug-in is loaded.

What is the predicate I should have used? `cond.wait({ thePluginController.howDoIKnowThatAPluginIsOpen })` -- I'd have expected `isOpen` or such. (With a little poking around, I find `.info.notNil` would do it, but I still think that VSTPluginController should provide a simple Boolean method for it. Users ideally shouldn't have to guess or explore.)

**Comment by smrg-lm** [comment]
> if you satisfy the condition and signal it before calling wait, it won't block.

This is *true* for `Condition.wait`, the problem if with `Condition.hang` which can be used as an unsafe `wait` because it doesn't check the test. `Condition.hang` may even fail in different ways, e.g. if a routine hangs with wait time, the time expires, the rescheduled routine goes to a next yield and `unhang`  or signal wakes it up before time because it wasn't removed from the queue.

There are some important things to consider with these pseudo threading stuff, the main language lock (which makes some things to be more deterministic), the local condition, that functions are scheduled and real threads are never blocked (because they are the scheduler), and maybe something else. It gets confusing.

**Comment by telephon** [comment]
> @Spacechild1 I see it will break if user code relies on it to block inside a routine,

sorry, can you explain a little?

I would have assumed that if I want the code to block, I need to wrap it in a Routine anyhow, because `loadToFloatArray` is an async concurrent operation. Or I would use the action callback to release a block that I have created myself externally.

**Comment by Spacechild1** [comment]
@telephon IIUC, it means that some user have been relying on the fact that `loadToFloatArray` and similar methods effectively block when executed inside a Routine (because they internally use `forkIfNeeded`). Of course, this behavior is not documented and can be considered a bug, but fixing it would break some (buggy) user code. That's a tough one...

**Comment by telephon** [comment]
>Of course, this behavior is not documented and can be considered a bug,

No it was intended – the reason why `forkIfNeeded` was used. But yes, it isn't documented.

Ah I see, the problem is that someone would them *not* expect it to block, because they want many of them in parallel. 
Wouldn't it perhaps enough to document this by saying that you have to wrap them in `fork { }`, which does express

**Comment by Spacechild1** [comment]
> No it was intended – the reason why forkIfNeeded was used. But yes, it isn't documented.

Ah! In this case the solution is very simple: keep everything as it is and improve the documentation :-)

> Wouldn't it perhaps enough to document this by saying that you have to wrap them in fork { } if you want them to run in parallel?

Yes!

**Comment by dyfer** [comment]
TL;DR I'm in the "it's a feature" camp :)

I definitely use `forkIfNeeded` relying on blocking the routine. For example I have code that does some async stuff (let's say loading buffers and starting synths). Now I want to put that code in a larger project that loads that code and relies on the synths being created before doing something else. If we didn't have `forkIfNeeded`, I'd need to use a condition and pass it to the module's code, signaling it at the end. If the module uses `forkIfNeeded`, it would just block that routine with no need for extra syncing mechanisms so I can simply run it inside a routine of the larger system and it will block it.

**Comment by elgiano** [comment]
I see a lot of information here (thank you all!) and I felt the need to summarize what's going on:

1. forkIfNeeded can kill the calling thread: 
the actual problem behind this is that CondVar can block indefinitely if `wait` is called without predicate (since forkIfNeeded runs synchronously inside fork in the OP, signalOne gets called before wait = wait forever)

2. forkIfNeeded is called unexpectedly by functions which are expected to fork, possibly leading to unexpected blocks:
we need to agree on a consistent approach and document it properly. Note that throughout SCClassLibrary, Routines are forked both by calling `fork` and `Routine {}` (actually there are instances of `Routine.run` and `Routine.new` as well)

3. CondVar vs Condition:
another topic, but it looks like we are in the middle of a transition, and we should figure out if we really need both classes

It would be nice to change the issue's title (or close it and create one or two more) once it's clearer where this is going :)

Now my 2c:
> Actually, it would be great if all asynchronous methods would use forkIfNeeded under the hood, because it would mean that we could write asynchronous code in a synchronous style just by putting everything in a Routine.

I agree, and especially for methods sending OSC messages, because if they are sent in a forked routine, it becomes impossible to capture them in OSC bundles with `Server::-makeBundle`.
Example:
```supercollider
fork {
        // here makeBundle exits before Synth is run, and captures nothing
	s.makeBundle(false) {
		fork { Synth(\default) }
	}.postln;
	// -> [ ]
	
	// here forkIfNeeded blocks, and Synth gets captured by makeBundle
	s.makeBundle(false) {
		forkIfNeeded { Synth(\default) }
	}.postln
	// -> [ [ 9, default, 1054, 0, 1, nil ] ]
}
```
I concretely encountered this problem recently while working on a UnitTest: Ndef creation uses MixedBundle, which calls Routine.new, making it "un-bundleable", and thus not "syncable"

**Comment by telephon** [comment]
This whole issue comes up as a problem again and again. Also JITLib is complicated because one has to pass bundles around and have a second set of messages that do not do this. It seems to me that the general pattern behind is that of an nested execution context.

Perhaps this can be generalised so that it captures all scheduling related things? It roughly behaves like nested scopes with defaults.

1. if no environment is given, create one
2. if it exists use this one and perhaps add some state to it
3. a closing environment passes control to the outer one
4. the last one closes the door (e.g. sends the bundle, runs the routine, renders the soundfile, ...)

Bigger thing, but when solved may simplify many things. 

Caveat: 
1. everything that should not be happening in that scope needs to be escaped (like `fork` within `forkIfNeeded`).
2. there may be very different execution context which may not work well together
3. ...?

**Comment by smrg-lm** [comment]
> we could write asynchronous code in a synchronous style just by putting everything in a Routine.

This is a really nice idea in general, I thought about it, one thing to consider is that there are still asynchronous commands within the server and have completion messages or reply when finished or like synthdef that can just use /sync message.

Another thought was that to mix any kind of message with events timing might fail in some cases if latency isn't enough  to compensate.

**Comment by dyfer** [comment]
In the interest of moving this forward, I'd like to summarize what I understand from the conversation so far:
1. the original issue was due to the use of `CondVar` without a predicate inside `forkIfNeeded`
  - should be solved by adding documentation covering this case
2. there is lack of documentation and lack of clarity for syncing mechanisms in SC (particularly `Condition` vs `CondVar`) 
  - as pointed out earlier, separate documentation/guide on this would be desirable
  - `loadToFloadArray's` documentation in particular should be updated to indicate that it uses `forkIfNeeded` and thus will block if called inside a routine
3. the use of `fork` and `forkIfNeeded` might not be consistent in the class library. IMO we should open a separate issue to discuss this.
4. `Condition` vs `CondVar`: AFAIR `CondVar` was designed with the idea that it might/will replace condition eventually. Given this example and the discussion that followed, I think it's an open question whether `Condition` should indeed be superseded/deprecated. In any case, we should start by implementing documentation (`2.` above) and revisit that topic again.

Finally, I think that the title of this issue is now inaccurate, in the light of the subsequent findings. Should we change it @jamshark70 ?

I think that the next actionable steps would be to make a PR(s) to address `1.` and `2.` above and possibly start another issue/thread about `3.`. What do you all think?

**Comment by jamshark70** [comment]
For points 1, 2, 3, all correct. (For 2, `getToFloatArray` will also block if the wait time is negative. Side note -- I think the implementation of negative wait time is incorrect, but it's not a big deal.)

For point 4, my opinion would be to move (gradually) in the direction of deprecating Condition.

I'll change the issue title.
